Mathlib/RingTheory/MvPolynomial/Symmetric.lean

```diff
@@ -309,4 +309,244 @@ theorem psum_isSymmetric (n : ℕ) : IsSymmetric (psum σ R n) := rename_psum _
+
+/-- The `n`th complete homogeneous symmetric `MvPolynomial σ R`. -/
+def hsymm (n : ℕ) : MvPolynomial σ R := ∑ μ : Exponent σ n, ∏ i : σ, X i ^ (μ.exp i)
+
+lemma hsum_def (n : ℕ) : hsymm σ R n = ∑ μ : Exponent σ n, ∏ i : σ, X i ^ (μ.exp i) := rfl
+
+@[simp]
+theorem hsymm_zero : hsymm σ R 0 = 1 := by
+  simp only [hsymm, pow_zero]
+  simp_rw [Exponent.zero_of_sum_eq_zero]
+  have (p : MvPolynomial σ R) : p ^ 0 = 1 := by simp
+  simp_rw [this]
+  have : ∏ i : σ, (1 : MvPolynomial σ R) = 1 := by simp
+  simp_rw [this]
+  have : ∑ x : Exponent σ 0, (1 : MvPolynomial σ R) = (Fintype.card (Exponent σ 0) : MvPolynomial σ R) := by simp only [Finset.sum_const,
+    nsmul_eq_mul, mul_one, Fintype.card_eq_sum_ones, smul_eq_mul]
+  rw [this, Exponent.exponentCard_zero]
+  exact Nat.cast_one
+  done
+
+@[simp]
+theorem hsymm_one : hsymm σ R 1 = ∑ i, X i := by
+  have (i : σ) (s : Exponent σ 1) : Exponent.exp s i = if (Equiv.ofBijective (Exponent.fintype_toExponent σ) ((Exponent.fintype_toExponent_bijective σ) : Function.Bijective (Exponent.fintype_toExponent σ))).symm s = i then 1 else 0 := by
+    have : ∃ (i : σ), Exponent.one σ i = s := ExistsUnique.exists (Exponent.one_sum_singleton σ s)
+    obtain ⟨j, h⟩ := this
+    have is_inv (j : σ) : (Equiv.ofBijective (Exponent.fintype_toExponent σ) ((Exponent.fintype_toExponent_bijective σ) : Function.Bijective (Exponent.fintype_toExponent σ))).symm (Exponent.one σ j) = j := by
+      have : (Equiv.ofBijective (Exponent.fintype_toExponent σ) ((Exponent.fintype_toExponent_bijective σ) : Function.Bijective (Exponent.fintype_toExponent σ))) j = Exponent.one σ j := by
+        simp [Exponent.fintype_toExponent, Exponent.one]
+      rw [←this]
+      simp only [Equiv.ofBijective_apply, Equiv.ofBijective_symm_apply_apply]
+    simp_rw [←h, is_inv j, Exponent.one]
+  have pow_eq (i x : σ) : (X x ^ if i = x then 1 else 0) = if i = x then ((X : σ → MvPolynomial σ R) i) else 1 := by
+    by_cases h : i = x
+    · simp_rw [if_pos h]
+      simp only [pow_one, h]
+    · simp_rw [if_neg h]
+      simp only [pow_zero]
+  calc
+    hsymm σ R 1 = ∑ μ : Exponent σ 1, ∏ i : σ, X i ^ Exponent.exp μ i := by simp only [hsymm, pow_one]
+    _ = ∑ i : σ, ∏ j : σ, X j ^ (Exponent.fintype_toExponent σ i).exp j := by apply Fintype.sum_equiv; swap; symm; exact Equiv.ofBijective _ (Exponent.fintype_toExponent_bijective σ); intro s; congr; simp only [Exponent.fintype_toExponent]; funext i; congr; simp only [this]
+    _ = ∑ i : σ, X i := by congr; funext i; simp only [Exponent.fintype_toExponent, Equiv.toFun_as_coe]; simp_rw [pow_eq _ _]; simp only [Finset.prod_ite_eq,
+      Finset.mem_univ, ite_true]
+
+theorem map_hsymm (n : ℕ) (f : R →+* S) : map f (hsymm σ R n) = hsymm σ S n := by
+  simp_rw [hsymm, map_sum, map_prod, map_pow, map_X]
+
+theorem rename_hsymm (n : ℕ) (e : σ ≃ τ) : rename e (hsymm σ R n) = hsymm τ R n :=
+  calc
+    rename e (hsymm σ R n) = ∑ μ : Exponent σ n, ∏ i : σ, X (e i) ^ (μ.exp i) := by
```

Comment:
It looks like you have multiple goals at some points. Please use focusing dots, to make sure only one goal is active at the time. (This also makes the proof nicer to read.)